### ISSUE TITLE
Support for "Lonsonho Tuya ZigBee Wireless Smart Scene Switch 4 Gang" needed

### ISSUE THREAD

**Issue by Spirituss** [Koenkk/zigbee2mqtt#4468]
## Information about the device + link
Lonsonho Tuya ZigBee Wireless Smart Scene Switch 4 Gang
![IMG_3893](https://user-images.githubusercontent.com/47206760/94346184-9b135680-0033-11eb-8a23-f793a224558d.jpg)

![IMG_3892](https://user-images.githubusercontent.com/47206760/94346190-a797af00-0033-11eb-97c7-71331c4e4d98.jpg)

![IMG_3894](https://user-images.githubusercontent.com/47206760/94346198-b1211700-0033-11eb-9f00-6a00ee805a38.jpg)

## data/database.db entry of the device
```
{"id":70,"type":"EndDevice","ieeeAddr":"0xbc.....","nwkAddr":...,"manufId":4098,"manufName":"_TZ3000_vp6clf9d","powerSource":"Battery","modelId":"TS0044","epList":[1,2,3,4],"endpoints":{"1":{"profId":260,"epId":1,"devId":0,"inClusterList":[0,10,1,6],"outClusterList":[25],"clusters":{"genBasic":{"attributes":{"65503":"^q�&\u0012","modelId":"TS0044","manufacturerName":"_TZ3000_vp6clf9d","powerSource":3,"zclVersion":3,"appVersion":66,"stackVersion":0,"hwVersion":1,"dateCode":""}},"genPowerCfg":{"attributes":{"batteryPercentageRemaining":200,"batteryVoltage":30}}},"binds":[]},"2":{"profId":260,"epId":2,"devId":0,"inClusterList":[1,6],"outClusterList":[],"clusters":{},"binds":[]},"3":{"profId":260,"epId":3,"devId":0,"inClusterList":[1,6],"outClusterList":[],"clusters":{},"binds":[]},"4":{"profId":260,"epId":4,"devId":0,"inClusterList":[1,6],"outClusterList":[],"clusters":{},"binds":[]}},"appVersion":66,"stackVersion":0,"hwVersion":1,"dateCode":"","zclVersion":3,"interviewCompleted":true,"meta":{},"lastSeen":1600948923006}
```

**Comment by lcse66** [comment]
Hi,
I have one of this.

It works with these changes:

folder /opt/zigbee2mqtt/node_modules/zigbee-herdsman-converters
```
--- devices_orig.js     2020-09-28 19:16:53.095887508 +0200
+++ devices.js  2020-09-28 19:16:19.846152157 +0200
@@ -1231,6 +1231,18 @@
         toZigbee: [],
     },
     {
+        zigbeeModel: ['TS0044'],
+        model: 'TS0044',
+        vendor: 'TuYa',
+        description: 'Wireless switch with 4 buttons',
+        whiteLabel: [
+            {vendor: 'Lonsonho', model: 'TS0044'},
+        ],
+        supports: 'action',
+        fromZigbee: [fz.tuya_on_off_action, fz.battery],
+        toZigbee: [],
+    },
+    {
         zigbeeModel: ['TS0001'],
         model: 'TS0001',
         vendor: 'TuYa',
```
folder /opt/zigbee2mqtt/node_modules/zigbee-herdsman-converters/converters
```
--- fromZigbee_orig.js  2020-09-28 19:36:08.504529036 +0200
+++ fromZigbee.js       2020-09-27 15:53:23.344632355 +0200
@@ -3521,6 +3521,7 @@
             let buttonMapping = null;
             if (model.model === 'TS0042') buttonMapping = {1: 'left', 2: 'right'};
             if (model.model === 'TS0043') buttonMapping = {1: 'right', 2: 'middle', 3: 'left'};
+            if (model.model === 'TS0044') buttonMapping = {1: 'button_1', 2: 'button_2', 3: 'button_3', 4: 'button_4'};
             const button = buttonMapping ? `${buttonMapping[msg.endpoint.ID]}_` : '';
             return {action: `${button}${clickMapping[msg.data[3]]}`};
         },
```

**Comment by Koenkk** [comment]
@lcse66 thanks, instead of `button_1`, can you change it to `left_top`, `right_bottom`, etc.? Then I can add support for it out of the box.

**Comment by lcse66** [comment]
```
--- fromZigbee_orig.js  2020-09-28 19:36:08.504529036 +0200
+++ fromZigbee.js       2020-09-27 15:53:23.344632355 +0200
@@ -3521,6 +3521,7 @@
             let buttonMapping = null;
             if (model.model === 'TS0042') buttonMapping = {1: 'left', 2: 'right'};
             if (model.model === 'TS0043') buttonMapping = {1: 'right', 2: 'middle', 3: 'left'};
+            if (model.model === 'TS0044') buttonMapping = {1: 'left_bottom', 2: 'right_bottom', 3: 'right_top', 4: 'left_top'};
             const button = buttonMapping ? `${buttonMapping[msg.endpoint.ID]}_` : '';
             return {action: `${button}${clickMapping[msg.data[3]]}`};
         },
```

**Comment by Koenkk** [comment]
Thanks, device will be supported in next Zigbee2mqtt release on 1 October.

**Comment by Spirituss** [comment]
I upgraded z2m to 115 and this devices still doesn’t work. I still can see device interview and announcement but there is no device in the device list.

**Comment by Koenkk** [comment]
@Spirituss can you provide me your Zigbee2MQTT startup log?

**Comment by Spirituss** [comment]
@Koenkk I cut lines with other devices.

[log.txt](https://github.com/Koenkk/zigbee2mqtt/files/5329640/log.txt)

**Comment by Koenkk** [comment]
@Spirituss did you set `zigbee_shepherd_devices: true` in your configuration? (this should be `false`).

**Comment by Spirituss** [comment]
@Koenkk I didn't set this option at all, should I do?

**Comment by VoyteckPL** [comment]
I just connected it and reaction to press is sometimes over 3-4 seconds. Why is that? Im on HA and Z2M 1.16.1.

**Comment by lcse66** [comment]
> I just connected it and reaction to press is sometimes over 3-4 seconds. 
It happens to me too. This delay is already visible on the serial device (/dev/ttyACM0, USB dongle based on CC2531, raspberry PI 3 B+), even with good link quality.

**Comment by VoyteckPL** [comment]
So it's software problem as it happends on conbee 2 and cc2531.

**Comment by VoyteckPL** [comment]
@Koenkk can we have this ticket reopened for futher investigation of latency?

**Comment by VoyteckPL** [comment]
Thanks, how can I investigate latency? I have other buttons (Aqara for example) and didn't experience big latency.

**Comment by Koenkk** [comment]
You can check by sniffing (https://www.zigbee2mqtt.io/how_tos/how_to_sniff_zigbee_traffic.html) , check how much time there is between clicking and seeing the message in wireshark.

**Comment by VoyteckPL** [comment]
That looks complicated... I have Conbee 2, Hassos 5.5 on RPI 4. Is there any simpler way?

**Comment by Koenkk** [comment]
I don't think (assuming your Conbee 2 is already connected through a USB extension cable I don't know what could cause this)

**Comment by jeroenkl** [comment]
I've got a ZZH! with TS0044 and not experiencing any delay.

**Comment by barmazu** [comment]
Actually I came here because of the matter you are now discussing. I've got ZZH! and I'm seeing quite delays with TS0044. This happens when I try to multiple actions fast - ex. do a "double" click after doing a "single" click (or vice versa) - the second and any other subsequent action will be delayed, not always but for most cases. If I run single action or allow device to "cool down" between actions for 2-3s it works w/o any delay. So this looks like TS0044 hardware fault rather than anything else?

**Comment by VoyteckPL** [comment]
> Actually I came here because of the matter you are now discussing. I've got ZZH! and I'm seeing quite delays with TS0044. This happens when I try to multiple actions fast - ex. do a "double" click after doing a "single" click (or vice versa) - the second and any other subsequent action will be delayed, not always but for most cases. If I run single action or allow device to "cool down" between actions for 2-3s it works w/o any delay. So this looks like TS0044 hardware fault rather than anything else?

I have exactly the same experience.

**Comment by jeroenkl** [comment]
I checked again and I do not have any delay. Clicking, single, double, hold subsequential (so many actions after in a row).
I ordered a second TS0044 and see if it is the same (or not)...
Or you running the latest ZZH! Koenkk firmware? Healty ZIgbee network?

**Comment by VoyteckPL** [comment]
his is weird. I'm on conbee 2, RPI4 4GB, SSD with USB 3.0 support. Small zigbee network (10 nodes) signal 100%. I get this delays too.

**Comment by barmazu** [comment]
> I checked again and I do not have any delay. Clicking, single, double, hold subsequential (so many actions in a row).
> I ordered a second TS0044 and see if it is the same (or not)...
> Are you running the latest ZZH! Koenkk firmware? Healty ZIgbee network?
Except TS0044 I have no issues within the network, all other zigbee buttons I have (apart TS0044 I've got 4 x Mi Wireless Zigbee and one Ikea Dimmer/Switch) work flawlessly, so the rest of devices - bulbs, temperature sensors, door/window sensors, wall switches, etc, ~35 devices in total. No drop outs, no delays. ZZH! is running 20200925 Koenkk's  Z-Stack firmware - rock solid, no plans for update at the moment.

**Comment by prestonsmartuk** [comment]
Can someone help me please.  Within the exposes of Zigbee2MQTT WebUI, it discovered it, recognised it was Tuya brand etc but it's showing this:
battery
Remaining battery in % N/A%
action
Triggered action (e.g. a button click) N/A

I can't use any of the buttons as they aren't showing up as useable in Home Assistant.  
Can someone help me fix please?

**Comment by Rincovanderbeek** [comment]
I have exactly the same, only the power status is visible as integration. No switches

**Comment by jeroenkl** [comment]
very strange, it is working fine on my site. No delay's (got 2 TS0044) and all the button's are visible.
I'm using the latest koenk version with Domoticz

**Comment by Spirituss** [comment]
For me it works too, but regarding delays - it takes place. Most of the time there is no delay, but sometimes I have significant (about 5 sec!!) delay for it.

**Comment by Rincovanderbeek** [comment]
Strange indeed; I use HA, latest version, zigbee integration and a Ziig-A-Zag USB flashed withCC26X2R1_20201026.
I tried it a couple of times, no succes.
Other devices are binding correctly.

**Comment by Rincovanderbeek** [comment]
Any idea why it only detects the powerstate?
Does it help when I list the device signing?

{
  "node_descriptor": "NodeDescriptor(byte1=2, byte2=64, mac_capability_flags=128, manufacturer_code=4098, maximum_buffer_size=82, maximum_incoming_transfer_size=82, server_mask=11264, maximum_outgoing_transfer_size=82, descriptor_capability_field=0)",
  "endpoints": {
    "1": {
      "profile_id": 260,
      "device_type": "0x0006",
      "in_clusters": [
        "0x0000",
        "0x0001",
        "0x0006",
        "0x000a"
      ],
      "out_clusters": [
        "0x0019"
      ]
    },
    "2": {
      "profile_id": 260,
      "device_type": "0x0006",
      "in_clusters": [
        "0x0001",
        "0x0006"
      ],
      "out_clusters": []
    },
    "3": {
      "profile_id": 260,
      "device_type": "0x0006",
      "in_clusters": [
        "0x0001",
        "0x0006"
      ],
      "out_clusters": []
    },
    "4": {
      "profile_id": 260,
      "device_type": "0x0006",
      "in_clusters": [
        "0x0001",
        "0x0006"
      ],
      "out_clusters": []
    }
  },
  "manufacturer": "_TZ3000_vp6clf9d",
  "model": "TS0044",
  "class": "zhaquirks.tuya.ts0044.TuyaSmartRemote0044"
}

**Comment by barmazu** [comment]
Finally I was able to run Zigbee sniff, and this TS0044 is a garbage, at least it looks so when comparing how Xiaomi switch/button (WXKG01LM)

After pressing a button on TS0044 3 zbee ZCL payload packets are sent:  First one  - a battery state attribute ("remaining battery power in %"). The second actual action payload - and action is performed INSTANTLY (eg light goes off). The weird thing is Wireshark is showing ZCL payload  as Malformed OnOff command, but nonetheless - it worked. And third zbee packet is shown around one second after the second one (this is the with empty action payload). In between there are Data Request packets sent by TS0044 where most major delays are seen (2x0.8s). What data is being requested? I have no idea. Now the tricky part: so imagine, if I press a button once again but somewhere in between, before 3rd zbee packet is processed and sent out by the device you will get a lag - at least with my device. Depending on how fast I have pressed the button again you will see higher or lower lag. So this second press will be processed only after first press of the button is fully completed (ie all 3 packets are sent out). How much does take to complete a "press a button action"? Around 2 - 2,5s. So this cool down thing I mentioned before make sense now (I mean it doesn't but you get the point). This is 100% reproducible, no matter what: how close to coordinator, sender receiver distance, firmware version (same for latest one as well as previous one). 

For mentioned Xiaomi it takes ~0.2-0.3s process and send action payload. No Data Requests packets are visible.
Under a cover I found a sticker which says: ESW-0ZAA-EU HRBG30LSE003C6
![obraz](https://user-images.githubusercontent.com/60849136/106392218-2beaac00-63f1-11eb-9f0d-38c4ed9adb09.png)

**Comment by Rincovanderbeek** [comment]
I just found out that the actions per button must be defined as automations.
So not as integrations. Not sure if that is normal but that is how it works with my device.

**Comment by dummys** [comment]
Hi, I have the same switch but I was not able to pair it with z2m. Are you able to pair it ? I opened the device and the MCU is a TUYA TYZS3 chip inside which should do Zigbee.

**Comment by Sjworks** [comment]
I'm having the same problem.
By the way, I found out that this button works well without any delay when linked with Smartthings Hub(with custom DTH https://github.com/YooSangBeom/SangBoyST/blob/master/devicetypes/sangboy/zemismart-button.src/zemismart-button.groovy ), and I tested a few things.

I tried sniffing packets respectively when connected to Zigbee2Mqtt and Smartthings.
All of the following situations are packets when two buttons are pressed at the same time.

With Smartthing Hub (filter by `zbee_newk`)
![4switch with st](https://user-images.githubusercontent.com/773003/107874497-6aa65900-6efd-11eb-8f5e-b27f450daa16.jpg)

With Zigbee 2 Mqtt (filter by `zbee_newk`)
![4switch with z2m](https://user-images.githubusercontent.com/773003/107874509-7abe3880-6efd-11eb-92f8-e48571070163.jpg)

I was able to see the `ZCL: Default Response` packet only in the Smartthings environment that is running normally,
In the `Z2M` environment, it seemed as if transmission failed and retryed several times.
I am not familiar with the packet structure of Zigbee, but there seems to be a delay due to a function that is not implemented.

[zigbee sniff.zip](https://github.com/Koenkk/zigbee2mqtt/files/5977560/zigbee.sniff.zip)

Is it possible to solve this problem?

**Comment by JulioC** [comment]
I was able to pair my Zemismart 4 gang Scene Switcher with the following devices.js definition (fingerprint taken from their [official specs](https://drive.google.com/file/d/1zQuF92Hs2ryj_nuzIPmwgKJB3o_rzFKL/view))

```js
    {
        zigbeeModel: ['TS0044'],
        model: 'TS0044',
        vendor: 'TuYa',
        description: 'Wireless switch with 4 buttons',
        whiteLabel: [{vendor: 'Lonsonho', model: 'TS0044'}, {vendor: 'Haozee', model: 'ESW-OZAA-EU'}],
        fromZigbee: [fz.tuya_on_off_action, fz.battery],
        exposes: [e.battery(), e.action(['1_single', '1_double', '1_hold', '2_single', '2_double', '2_hold',
            '3_single', '3_double', '3_hold', '4_single', '4_double', '4_hold'])],
        toZigbee: [],
        fingerprint: [
            {manufacturerName: "_TZ3400_key8kk7r", modelID: "TS0043", },
            {manufacturerName: "_TYZB02_key8kk7r", modelID: "TS0043", },
            {manufacturerName: "_TZ3000_bi6lpsew", modelID: "TS0043", },
            {manufacturerName: "_TZ3000_vp6clf9d", modelID: "TS0044", }
        ]
    }
```

**Comment by matlab22** [comment]
> 
> 
> Hi, I have the same switch but I was not able to pair it with z2m. Are you able to pair it ? I opened the device and the MCU is a TUYA TYZS3 chip inside which should do Zigbee.

I have the same problem. All 4 LEDs are blinking, but it does not show up in the info log :(

**Comment by rensenti** [comment]
> I am not familiar with the packet structure of Zigbee, but there seems to be a delay due to a function that is not implemented.
> 
> [zigbee sniff.zip](https://github.com/Koenkk/zigbee2mqtt/files/5977560/zigbee.sniff.zip)
> 
> Is it possible to solve this problem?

Has anyone had a chance to investigate this? The delay is quite annoying for my particular use-case, remote(s) for roller shutter blinds.

What can we do to help?

**Comment by Ramblurr** [comment]
I'm interested in a multi button Zigbee switch like this and i have zzh! 

Over in ZHA land they have an additional module/library called zha-quirks that seems to fix this switch  

https://community.home-assistant.io/t/tuya-4-gang-scene-switch/264164/9

**Comment by kuradi** [comment]
i really dont understand now... https://www.zigbee2mqtt.io/information/supported_devices.html saying this switch is supported but i cant get it working. i got "device announced" but when i push switch log shows: {"level":"warning","message":"Please see: https://www.zigbee2mqtt.io/how_tos/how_to_support_new_devices.html."}. please advise.

**Comment by VoyteckPL** [comment]
Is it just me or anyone else experience delay with this switch? Sometimes if I press it it takes 1-1,5 seconds before it reacts.

**Comment by dummys** [comment]
> Is it just me or anyone else experience delay with this switch? Sometimes if I press it it takes 1-1,5 seconds before it reacts.

same here, never was able to fix

**Comment by HB9VCT** [comment]
I have to confirm that TS0043 has a random delays up to 10 seconds with Z2M. After sniffing the protocol I have made the conclusion that this device is waiting for some kind of acknowledge message. Since it is not seen it repeat the command several times and is blocked during this time.
Should be nice is somebody (Koenkk) can have a look at it.

**Comment by Sjworks** [comment]
@dummys @VoyteckPL @rensenti @Spirituss 
The delay problem is being solved with Koenkk's help on #8149..
Please check comments in there.
